backend/src/connector/mod.rs

```diff
@@ -1 +1,267 @@
+        hyper::client::connect::HttpConnector::new()
+    }
+}
+
+// TODO
+/// [Stub] Builder for HTTPS connectors
+#[derive(Debug)]
+pub struct HttpsBuilder {}
+
+impl HttpsBuilder {
+    pub fn build(self) {
+        unimplemented!()
+    }
+}
+
+#[derive(Clone)]
+pub struct BobClient {
+    /// Bob's hostname
+    hostname: Hostname,
+
+    // NOTE: Can (and should) the API interface mutate?..
+    /// Connection,
+    main: Arc<ApiInterface>,
+
+    /// Clients for all known nodes
+    cluster: HashMap<NodeName, Arc<ApiInterface>>,
+}
+
+#[allow(clippy::missing_fields_in_debug)]
+impl std::fmt::Debug for BobClient {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let user = &self
+            .context()
+            .get::<Option<Authorization<Basic>>, _>()
+            .as_ref()
+            .map_or("Unknown", |cred| cred.username());
+        f.debug_struct("BobClient")
+            .field("hostname", &self.hostname)
+            .field("user", &user)
+            .finish()
+    }
+}
+
+impl BobClient {
+    /// Creates new [`BobClient`] from [`BobConnectionData`]
+    ///
+    /// # Errors
+    /// The function will fail if a hostname isn't a valid url or the client couldn't establish
+    /// connection for the BOB cluster
+    pub async fn try_new(
+        bob_data: BobConnectionData,
+        timeout: RequestTimeout,
+    ) -> Result<Self, ClientError> {
+        let auth = if let Some(creds) = bob_data.credentials {
+            Some(Authorization::basic(&creds.login, &creds.password))
+        } else {
+            None
+        };
+        let hostname = bob_data.hostname.clone();
+
+        let context: ClientContext = hlist![timeout, auth, XSpanIdString::default()];
+        let client = Box::new(
+            Client::try_new_http(&hostname.to_string()).change_context(ClientError::InitClient)?,
+        );
+        let nodes = client
+            .with_context(context.clone())
+            .get_nodes()
+            .await
+            .change_context(ClientError::InitClient)?;
+        let api::GetNodesResponse::AJSONArrayOfNodesInfoAndVdisksOnThem(nodes) = nodes else {
+            Err(APIError::InvalidStatusCode(StatusCode::FORBIDDEN))
+                .change_context(ClientError::PermissionDenied)?
+        };
+
+        let cluster: HashMap<NodeName, Arc<ApiInterface>> = nodes
+            .iter()
+            .map(|node| {
+                (
+                    &node.name,
+                    Self::change_port(&node.address, &bob_data.hostname),
+                )
+            })
+            .filter_map(|(name, hostname)| {
+                if hostname.is_err() {
+                    tracing::warn!("couldn't change port for {name}. Client won't be created");
+                }
+                hostname
+                    .ok()
+                    .map(|hostname| (name, Client::try_new_http(&hostname.to_string())))
+            })
+            .filter_map(|(name, client)| {
```

Comment:
Don't see a reason to chain multiple `map` and `filter_map`. Everything can be done within the single `filter_map` call and it seems to me, that the code become simpler